src/docfx/lib/git/GitUtil.cs

```diff
@@ -0,0 +1,360 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Docs.Git;
+
+using FileMap = System.Collections.Generic.Dictionary<string, (string file, string parent)>;
+using TreeMap = System.Collections.Concurrent.ConcurrentDictionary<long, System.Collections.Generic.Dictionary<string, long>>;
+
+namespace Microsoft.Docs
+{
+    /// <summary>
+    /// Provide git operations
+    /// </summary>
+    public static class GitUtil
+    {
+        /// <summary>
+        /// Find git repo directory
+        /// </summary>
+        /// <param name="path">The git repo entry point</param>
+        /// <returns>The git repo root path</returns>
+        public static string FindRepo(string path)
+        {
+            var repo = path;
+            while (!string.IsNullOrEmpty(repo))
+            {
+                var gitPath = Path.Combine(repo, ".git");
+                if (Directory.Exists(gitPath) || File.Exists(gitPath) /* submodule */)
+                {
+                    return repo;
+                }
+                repo = Path.GetDirectoryName(repo);
+            }
+            return repo;
+        }
+
+        /// <summary>
+        /// Get git repo information, will change to git exe later
```

Comment:
>will change to git exe later [](start = 38, length = 28)

Why don't we do it right now? I don't think we need to rush.

Reply:
we don't need it right now, i will remove it from this pr

---
In reply to: [179996870](https://github.com/dotnet/docfx/pull/2601#discussion_r179996870) [](ancestors = 179996870)